httpx/models.py

```diff
@@ -613,7 +614,32 @@ def prepare(self) -> None:
         has_connection = "connection" in self.headers
 
         if not has_host:
-            auto_headers.append((b"host", self.url.authority.encode("ascii")))
+            _params: typing.Dict[str, None] = {}
+            if self.url._uri_reference.userinfo:
```

Comment:
We shouldn't need to reach into the private `_uri_reference` from anywhere outside of the `URL` class. If we're doing that anywhere then we ought to treat it as a signal that we're missing some required bit of interface on the `URL` class.

Reply:
yes but I didn't want to perform it every time regarding performance.

Reply:
Gotcha, so...

* We should probably add a `.userinfo` property to the `URL` class.
* We should then use something like...

```python
url = self.url
if url.userinfo:
    url = url.copy_with(username=None, password=None)
...
```

Right?

Reply:
to copy `userinfo`, added `copy_with_urlparse` Is it appropriate?